### PR TITLE
[MRG] fix logic checking extinput synweights zero

### LINKED ISSUE #101: alpha simulation fails (NEURON?)
Walkthroughs in `examples` for `evoked` and `gamma` work fine, but [`alpha`](https://github.com/jonescompneurolab/hnn-core/blob/master/examples/plot_simulate_alpha.py) fails mysteriously. I've set `'tstop': 110.0` for speed, but below is equally valid for the default `710.0`.

 

In the hopes of finding a bug in `params`, I even tried this (3 keys differ, not that they seemed important):

```python
params_fname = op.join(hnn_core_root, '../hnn/param', 'OnlyRhythmicDist.param')
```

But I get the same plots. The same example works fine in `hnn`, in the same `conda` env., where NEURON is from PyPi (`7.8.0-127-g91f6e623+`).

I guess what I'm asking is: any pointers on where to start debugging my setup? I suppose it's likely my NEURON installation, though I've been through the tests for `nrnpython`, which run quite happily in a parallel context (`openmpi`, `mpi4py`) on my dual-core macOS Catalina laptop.

### PR DESCRIPTION
**Pull request by cjayb**
This PR fixes the part of #98 (specifically 94fc21a6 in `feed.py`) that botches the alpha-example.

No test written yet, input appreciated! There's a refactor brewing (see also discussion in #98), so perhaps tests should be deferred? To speed up tests, would it make sense to create a 'minimal network'? Not sure that makes sense, but almost 3 min spent on just the full alpha-example isn't sustainable. 

Fixes #101

### CODE REVIEW

**Comment by jasmainak** [comment]
@rythorpe do you want to review this PR since the commit that introduces the problem was yours?

File: hnn_core/feed.py [hunk 1]
```diff
@@ -180,13 +180,17 @@ def __create_extinput(self):
...
             if key.startswith('L2Pyr') or \
                     key.startswith('L5Pyr') or \
                     key.startswith('L2Bask') or \
                     key.startswith('L5Bask'):
-                if self.p_ext[key][0] <= 0.0:
-                    return False
+                if self.p_ext[key][0] > 0.0:
+                    all_syn_weights_zero = False
+        if all_syn_weights_zero:
+            return False
```
Comment (jasmainak) [inline hnn_core/feed.py:192]: it's a bit weird that this method even returns something because the return value is not used anywhere

Comment (rythorpe) [reply hnn_core/feed.py:192]: @cjayb I see, so the error in #101 is caused by one of the synaptic weights from 'OnlyRhythmicDist.param' being zero, thus causing `__create_extinput()` to return `False` even though some of the other weights are non-zero?

Comment (rythorpe) [reply hnn_core/feed.py:192]: To address @jasmainak's point, this might be a good time to rethink the return values of the `__create_` methods in the `ExtFeed` class. It looks like the code is currently set up to return the number of inputs added to the feed (for troubleshooting purposes?), though this might be unnecessary if we implement good tests.

Comment (jasmainak) [reply hnn_core/feed.py:192]: yes, +10 for tests! @rythorpe do you have a suggestion for a test based on your experience in #101 ?

Comment (cjayb) [reply hnn_core/feed.py:192]: > so the error in #101 is caused by one of the synaptic weights from 'OnlyRhythmicDist.param' being zero

Correct, these are the only non-zero weights:

```python
    'input_dist_A_weight_L2Pyr_ampa': 5.4e-5,
    'input_dist_A_weight_L5Pyr_ampa': 5.4e-5,
```

The method would return unless _all_ weights were non-zero.

> rethink the return values of the `__create_` methods in the `ExtFeed` class

+1 It doesn't seem that the return values are ever used

Comment (cjayb) [reply hnn_core/feed.py:192]: > if we implement good tests

Right :) So how do we do this? I can update this PR to remove the return arguments, and replace with a test. But how?

As I mentioned, running the full alpha example takes too long. We could use a smaller `'tstop'`, then do something like in #99 for the `spikegids` of type `'extinput'`?

**Comment by cjayb** [comment]
Re: tests, I just did

```python
    params_fname = op.join(hnn_core_root, 'param', 'default.json')
    params = read_params(params_fname)
    params.update({'input_dist_A_weight_L2Pyr_ampa': 5.4e-5,
                                 'input_dist_A_weight_L5Pyr_ampa': 5.4e-5,
                                 't0_input_dist': 50})
...    
```

which adds `'extinput': 33,` to `spiketype_counts`. The assertions against `dpl_master` will fail, but I'm pretty certain this would have caught the bug in #101.

Still don't like the length of this test, though... Just for the heck of it, I tried

```python
params.update({"N_pyr_x": 2, "N_pyr_y": 2})
```

and re-run `simulate_dipole`. The model compiled and ran in just a few seconds, and produced some output

```python
spiketype_counts
{'extinput': 34,
 'evprox1': 11,
 'L5_basket': 5,
 'L2_pyramidal': 13,
 'L2_basket': 6,
 'L5_pyramidal': 31,
 'evdist1': 9,
 'evprox2': 11}
```

Could this be a strategy? Shall we close this PR, then I try to create a fast test _a la_ the above in a separate one?

**Comment by jasmainak** [comment]
@cjayb I like the idea of having tests on a small network! However, I'm not sure if the signals you get out of the smaller network will be as meaningful to know if we broke something?

Re the old tests of comparing to `dpl.txt`, I would not touch it yet. I think once we have CircleCI running in #97 we can automatically check the plots like in MNE-Python for all the examples. But we need more tests which look like "unit" tests. That is, they check small parts of the code.

File: hnn_core/feed.py (hunk 1 above)
Comment (jasmainak) [reply hnn_core/feed.py:192]: I think all these are great ideas for faster tests. I'm still having trouble understanding what exactly you fixed though. It seems to me that all the logic in the diff is modifying the return value but the return value is not used anywhere? What exactly changed from `master`? Maybe a good next step would be to update `whats_new.rst` so we write a description of the fix in plain english!

Comment (cjayb) [reply hnn_core/feed.py:192]: Follow the loop logic: it should only end the function if _all_ inputs are zero. Before it would end if _any_ input was <=0.
I’ll update the rst, but I think this was an unintentional mod (bug)

**Comment by cjayb** [comment]
My thought was that all the simulations are deterministic, so even though they would be physiologically meaningless, they could serve as unit-like tests? Given that most of what we do involves nrn, we’re going to have to deal with the long compile/run times somehow...

File: hnn_core/feed.py (hunk 1 above)
Comment (jasmainak) [reply hnn_core/feed.py:192]: okayyyy, now I get it

so all you need to do for the unit test is to initialize an `ExtFeed` object with the right parameters and then check if `self.eventvec` exists (or is not None. I didn't check exactly what this function does). No computation required at all.

Comment (jasmainak) [reply hnn_core/feed.py:192]: > I’ll update the rst, but I think this was an unintentional mod (bug)

yep it should go to the bug section of `whats_new.rst`

Comment (rythorpe) [reply hnn_core/feed.py:192]: Continuing off @jasmainak's suggestion, ideally we'd check the number and type of inputs in the `p_ext` attribute of an `ExtFeed` object so that we can be sure all (vs. any) expected inputs were added to the feed. I think this test could be added to the network test script.

**Comment by jasmainak** [comment]
> My thought was that all the simulations are deterministic, so even though they would be physiologically meaningless, they could serve as unit-like tests?

But that's not a very good unit test because if you fix a bug that changes the overall output, your test will fail. This is the case with the test at the moment. The existing test is a *bad* unit test. It's just better than nothing.

File: hnn_core/feed.py (hunk 1 above)
Comment (cjayb) [reply hnn_core/feed.py:192]: see my latest push: it wasn't enough to check `p_ext`, since these could be copied in fine, despite `__create_extinput` not running through to create the `eventvec`. I think we should test `p_ext` in `create_pext`?

**Comment by cjayb** [comment]
> But that's not a very good unit test because if you fix a bug that changes the overall output, your test will fail.

OK, got it.

File: hnn_core/tests/test_network.py [hunk 2]
```diff
@@ -28,3 +29,35 @@ def test_network():
     for ev_input in params['t_ev*']:
         type_key = ev_input[2: -2] + ev_input[-1]
         assert len(net.gid_dict[type_key]) == net.N_cells
+
+
+def test_external_rhythmic_feeds():
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:34]: I would actually put this in a new file called `test_feed.py`

File: hnn_core/tests/test_network.py [hunk 3]
```diff
@@ -28,3 +29,35 @@ def test_network():
...
+    # from Network.build: creates ExtFeeds in < 1 sec
+    net._create_all_src()
+
+    # annoyingly, net.extinput_list is always 2, whether populated
+    # or not, so the first few assertions have no effect
+    # the last one needs to be > 0, though (catches #101, tests #102)
+    # Assert number of rhythmic inputs is 2 (distal & proximal)
+    assert len(net.extinput_list) == 2
+    for ei in net.extinput_list:
+        # naming could be better
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:58]: yes this should be called 'rhythmic'. We've discussed this with @rythorpe !

File: hnn_core/tests/test_network.py (hunk 2 above)
Comment (cjayb) [reply hnn_core/tests/test_network.py]: done

**Comment by jasmainak** [comment]
@cjayb I think your PR looks good to go for me. I played around with it a bit and removed some comments. I hope that's okay for you? I prefer to see clean code through good choice of variable names than explaining it through comments :)

If you're fine, please set the PR title to [MRG] and I'll merge in the morning when I wake up. Thanks a lot for the PR. Looking forward to more contributions in the future! Maybe an easy one is to change the naming of 'extinput' -> 'rhythmic'?

**Comment by cjayb** [comment]
Nice touch with the context manager! I definitely don't mind the edits. Good way to work this: add copious comments while working on PR, remove most once test is written.

Sure, I'll take a stab at the renaming next. Goal will be to create some test(s) while I'm at it.

**Comment by jasmainak** [comment]
Perfect plan! Thanks a lot :-)
